### PR TITLE
Integrate goto-synthesizer into Kani

### LINKED ISSUE #2156: Intergrate ```goto-synthesizer``` into Kani
The support for Kani-compiled GOTOs is implemented in https://github.com/diffblue/cbmc/pull/7487 and https://github.com/diffblue/cbmc/pull/7488. Once both of them have been merged. We will integrate ```goto-synthesizer``` into Kani so that users can choose to synthesize loop invariants for loop with a new flag.

### PR DESCRIPTION
**Pull request by qinheping**
### Description of changes: 

This PR integrate ```goto-synthesizer``` into Kani, and add an option to synthesize loop contracts for all loops.

### Resolved issues:

Resolves #2156 

### Call-outs:

The purpose of this PR is to allow us to experiment the loop-contracts synthesizer on Kani benchmarks. However, there are still a few issues we need to address with later PRs:

- The synthesizer may not terminate. When there is no loop invariants in the candidate space, the enumerator in the synthesizer will run forever trying to find a solution. To address this problem, we might want to allow users to limit the resource used by the synthesizer. On the ```goto-synthesizer``` side, we will add an option to limit the size of candidates we enumerate, so that it will not keep enumerating when there is no solution in the candidate space. On the Kani side, we can also add an option to limit the running time of the synthesizer.
- The synthesizer is not compatible with unwinding numbers. Now the synthesizer ignores unwinding numbers, and will synthesize loop-contracts for all loops no mater if there are loop numbers specified for them. However, we might sometimes want to unwind some loops while synthesize loop contracts for other loops. We will allow users to specify unwinding numbers in ```goto-synthesizer```. Then users can choose to synthesize loop contracts only for those loops without unwinding number specified.

### Testing:

* How is this change tested?

This PR include a regression test.

* Is this a refactor change?

No.

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-driver/src/harness_runner.rs
Comment (celinval) [inline kani-driver/src/harness_runner.rs:69]: How come?

File: kani-driver/src/harness_runner.rs
Comment (celinval) [inline kani-driver/src/harness_runner.rs:68]: Please check if we are on quiet mode before printing this message.

**Review by celinval — COMMENTED** [review]
That's awesome! Thanks for adding this. I have a few questions:
1. How does the loop synthesizer interacts with loop unwind argument?
2. Is there anyway we can avoid the non-termination? Even if that means failing the synthesis?
3. We need to add some documentation around this. I would suggest writing an RFC explaining the user experience, the value it adds and and its limitations.

**Comment by qinheping** [comment]
> 

Yes, in the Call-outs part of the PR description I introduced my plan to handle non-termination and how to deal with unwinding numbers. More comments about them are appropriated. 

And yes, I will create a RFC about it.

**Comment by celinval** [comment]
I think an RFC will help. From what you described, there is a conflict between unwind and loop synthesis, at least for now. In this case, we should figure out what the message should be and how we handle a harness that is annotated with `unwind`. There's also a question whether this should be applied to the entire system or to specific harnesses.

**Comment by zhassan-aws** [comment]
Is it possible to add tests?

**Comment by qinheping** [comment]
> Is it possible to add tests?

I added two simple tests to check the synthesizer is correctly called and enumerates required loop invariants. More experiments are needed to see how the synthesizer works on more complicated tests.

File: kani-driver/src/harness_runner.rs
Comment (qinheping) [reply kani-driver/src/harness_runner.rs]: Add

File: kani-driver/src/call_goto_synthesizer.rs
Comment (celinval) [inline kani-driver/src/call_goto_synthesizer.rs:2]: please add a little doc to this file

File: kani-driver/src/call_goto_synthesizer.rs
Comment (celinval) [inline kani-driver/src/call_goto_synthesizer.rs:24]: Is there a reason why you have this?

File: kani-driver/src/call_goto_synthesizer.rs
Comment (celinval) [inline kani-driver/src/call_goto_synthesizer.rs:12]: doc please

File: kani-driver/src/harness_runner.rs
Comment (celinval) [inline kani-driver/src/harness_runner.rs:68]: please move this code to be inside the synthesize_loop_contracts function.

File: kani-driver/src/harness_runner.rs
Comment (celinval) [inline kani-driver/src/harness_runner.rs:70]: please use `util::warn()`

File: tests/kani/LoopContractsSynthesizer/main_signed.rs
Comment (celinval) [inline tests/kani/LoopContractsSynthesizer/main_signed.rs:14]: can you please move these tests to the `ui` suite and ensure that the last assert is reachable.

File: kani-driver/src/args.rs
Comment (celinval) [inline kani-driver/src/args.rs:228]: you should add that this conflicts with `--unwind`

**Comment by qinheping** [comment]
> 

Thank you Celina! Yes, I plan to add more tests. I am doing the experiment of running the synthesizer on all current Kani benchmarks, which will provide us better insight about what we can and cannot synthesize. I will create an issue and add tests when the experiment is done.

File: tests/kani/LoopContractsSynthesizer/main_signed.rs
Comment (qinheping) [reply tests/kani/LoopContractsSynthesizer/main_signed.rs]: I moved them to the ```ui``` suite. I tried to use ```kani::cover``` instead of ```assert!```, but the synthesizer fail to solve the problem. I will create an issue about it and fix this issue.

File: kani-driver/src/call_goto_synthesizer.rs
Comment (zhassan-aws) [inline kani-driver/src/call_goto_synthesizer.rs:16]:
```suggestion
    /// used to prove some assertions, and applies found invariants when all checks pass.
```
?

File: kani-driver/src/call_goto_synthesizer.rs
Comment (zhassan-aws) [inline kani-driver/src/call_goto_synthesizer.rs:19]: I would say "Running loop contract synthesizer".

Comment (zhassan-aws) [inline kani-driver/src/call_goto_synthesizer.rs:19]: Does this need to be a warning?

File: kani-driver/src/call_goto_synthesizer.rs
Comment (zhassan-aws) [inline kani-driver/src/call_goto_synthesizer.rs:32]: Does the `goto-synthesizer` produce any output while it's running? If so, `run_terminal` might be more suitable.

File: tests/ui/LoopContractsSynthesizer/main_signed/expected
Comment (zhassan-aws) [inline tests/ui/LoopContractsSynthesizer/main_signed/expected:1]: The test file itself is missing from the PR.

File: tests/ui/LoopContractsSynthesizer/main_unsigned/expected
Comment (zhassan-aws) [inline tests/ui/LoopContractsSynthesizer/main_unsigned/expected:1]: Same for this one.

**Review by zhassan-aws — COMMENTED** [review]
Thanks @qinheping!

File: kani-driver/src/call_goto_synthesizer.rs
Comment (qinheping) [reply kani-driver/src/call_goto_synthesizer.rs:19]: Not actually, I will use ```println!``` instead.

File: kani-driver/src/call_goto_synthesizer.rs
Comment (qinheping) [reply kani-driver/src/call_goto_synthesizer.rs:32]: It only output debug messages when a high ```verbosity``` is specified.

Comment (zhassan-aws) [reply kani-driver/src/call_goto_synthesizer.rs:32]: I see. In this case, I suggest doing:
```rust
       if self.args.quiet {
            self.run_suppress(cmd)?;
        } else {
            self.run_terminal(cmd)?;
        }
```
so that we can see its output if needed.

Comment (qinheping) [reply kani-driver/src/call_goto_synthesizer.rs:32]: Sounds great!

**Review by zhassan-aws — APPROVED** [review]
Thanks @qinheping! Please address any remaining comments from @celinval and wait for her approval before merging.

File: kani-driver/src/call_goto_synthesizer.rs
Comment (celinval) [inline kani-driver/src/call_goto_synthesizer.rs:36]: You don't need this. In fact, `run_suppress` invokes `run_terminal` if `quiet` is true.

File: kani-driver/src/call_goto_synthesizer.rs
Comment (celinval) [inline kani-driver/src/call_goto_synthesizer.rs:20]: the `warning` function already adds the `warning` prefix.

File: kani-driver/src/args.rs
Comment (celinval) [inline kani-driver/src/args.rs:572]: I'm OK with the warning if the user has sets `default-unwind` but everything else I would highly recommend an error. I prefer if we don't have an implicit precedence for our arguments. If they are incompatible, just throw an error.

File: kani-driver/src/call_goto_synthesizer.rs
Comment (zhassan-aws) [reply kani-driver/src/call_goto_synthesizer.rs:36]: My bad (this was my suggestion) :(

Comment (zhassan-aws) [reply kani-driver/src/call_goto_synthesizer.rs:36]: We should clean up this instance then: https://github.com/model-checking/kani/blob/035225ca362b0a80c807af0f567c3513237cb0ee/kani-driver/src/concrete_playback.rs#L191

**Pull request merged**
